src/gc/gcinterface.h

```diff
@@ -0,0 +1,394 @@
+};
+
+// !!!!!!!!!!!!!!!!!!!!!!!
+// make sure you change the def in bcl\system\gc.cs 
+// if you change this!
+enum start_no_gc_region_status
+{
+    start_no_gc_success = 0,
+    start_no_gc_no_memory = 1,
+    start_no_gc_too_large = 2,
+    start_no_gc_in_progress = 3
+};
+
+enum end_no_gc_region_status
+{
+    end_no_gc_success = 0,
+    end_no_gc_not_in_progress = 1,
+    end_no_gc_induced = 2,
+    end_no_gc_alloc_exceeded = 3
+};
+
+typedef BOOL (* walk_fn)(Object*, void*);
+typedef void (* gen_walk_fn)(void *context, int generation, uint8_t *range_start, uint8_t * range_end, uint8_t *range_reserved);
+
+// IGCHeap is the interface that the VM will use when interacting with the GC.
+class IGCHeap {
+public:
+    virtual HRESULT Initialize() = 0;
+    virtual BOOL IsPromoted(Object *object) = 0;
+    virtual BOOL IsHeapPointer(void* object, BOOL small_heap_only = FALSE) = 0;
+    virtual Object* GetNextFinalizable() = 0;
+    virtual unsigned WhichGeneration(Object* obj) = 0;
+    virtual size_t GetTotalBytesInUse() = 0;
+    virtual unsigned GetCondemnedGeneration() = 0;
+    virtual int GetGcLatencyMode() = 0;
+    virtual int SetGcLatencyMode(int newLatencyMode) = 0;
+    virtual int GetLOHCompactionMode() = 0;
+    virtual void SetLOHCompactionMode(int newLOHCompactionMode) = 0;
+    virtual BOOL RegisterForFullGCNotification(uint32_t gen2Percentage, uint32_t lohPercentage) = 0;
+    virtual BOOL CancelFullGCNotification() = 0;
+    virtual int WaitForFullGCApproach(int millisecondsTimeout) = 0;
+    virtual int WaitForFullGCComplete(int millisecondsTimeout) = 0;
+    virtual int CollectionCount(int generation, int get_bgc_fgc_coutn = 0) = 0;
+    virtual size_t GetLastGCStartTime(int generation) = 0;
+    virtual size_t GetLastGCDuration(int generation) = 0;
+    virtual size_t GetNow() = 0;
+
+    // TODO(segilles) these are all referenced by the VM. Are all of them
+    // essential?
+    virtual BOOL    IsGCInProgressHelper (BOOL bConsiderGCStart = FALSE) = 0;
+    virtual unsigned GetMaxGeneration() = 0;
+    virtual void SetCardsAfterBulkCopy(Object** obj, size_t length) = 0;
+    virtual BOOL IsValidSegmentSize(size_t size) = 0;
+    virtual BOOL IsValidGen0MaxSize(size_t size) = 0;
+    virtual HRESULT GarbageCollect(int generation = -1, BOOL low_memory_p = FALSE, int mode = collection_blocking) = 0;
+    virtual void WaitUntilConcurrentGCComplete() = 0;
+    virtual BOOL IsConcurrentGCInProgress() = 0;
+    virtual size_t GetValidSegmentSize(BOOL large_seg = FALSE) = 0;
+    virtual BOOL FinalizeAppDomain(AppDomain *pDomain, BOOL fRunFinalizers) = 0;
+    virtual void SetFinalizeQueueForShutdown(BOOL fHasLock) = 0;
+    virtual size_t GetNumberOfFinalizable() = 0;
+    virtual BOOL ShouldRestartFinalizerWatchDog() = 0;
+    virtual unsigned GetGcCount() = 0;
+    virtual bool IsThreadUsingAllocationContextHeap(gc_alloc_context* acontext, int thread_number) = 0;
+    virtual BOOL IsObjectInFixedHeap(Object *pObj) = 0;
```

Comment:
this should only be used when VERIFY_HEAP is defined. are you seeing an exception to this?